### PR TITLE
Finish Rpc Framework and create sample method "CreateConnection"

### PR DESCRIPTION
**Pull request by realwanpengli**
1. Rpc framework dynamically creates methods by reflection.
2. Rpc slave can send exceptions to master.
3. Put two dictionaries for master and slave to store some variables which may be used across steps.

@vwxyzh @ccic please take a look at the design and structure of the projects in `azure-signalr-bench/SignalRServiceBenchmarkPlugin/` directory to see if there is something bad, thank you.

I plan to implement methods (echo, broadcast, etc) after these commit if everything is okay.

### CODE REVIEW

File: SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/MasterMethods/CreateConnection.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace Plugin.Microsoft.Azure.SignalR.Benchmark.MasterMethod
+{
+    public class CreateConnection : IMasterMethod
+    {
+        public Task Do(IDictionary<string, object> stepParameters, IDictionary<string, object> pluginParameters, IList<IRpcClient> clients)
+        {
+            var success = true;
+
+            success = stepParameters.TryGetTypedValue(SignalRConstants.ConnectionTotal, out int connectionTotal, Convert.ToInt32);
```
Comment (ccic) [inline SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/MasterMethods/CreateConnection.cs]: Is it possible to move PluginUtils.HandleGetValueResult into TryGetTypedValue? That method seems to be part of HandleGetValueResult.

Comment (realwanpengli) [reply SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/MasterMethods/CreateConnection.cs]: Make sense, thanks

File: SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/SlaveMethods/CreateConnection.cs [hunk 2]
```diff
@@ -24,25 +24,12 @@ public async Task Do(IDictionary<string, object> stepParameters, IDictionary<str
...
-
-                success = stepParameters.TryGetTypedValue(SignalRConstants.HubProtocol, out string protocol, Convert.ToString);
-                PluginUtils.HandleGetValueResult(success, SignalRConstants.HubProtocol);
-
-                success = stepParameters.TryGetTypedValue(SignalRConstants.TransportType, out string transportType, Convert.ToString);
-                PluginUtils.HandleGetValueResult(success, SignalRConstants.TransportType);
-
-                success = stepParameters.TryGetTypedValue(SignalRConstants.Type, out string type, Convert.ToString);
-                PluginUtils.HandleGetValueResult(success, SignalRConstants.Type);
+                PluginUtils.TryGetTypedValue(stepParameters, SignalRConstants.ConnectionBegin, out int connectionBegin, Convert.ToInt32);
```
Comment (realwanpengli) [inline SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/SlaveMethods/CreateConnection.cs]: @ccic  Moved PluginUtils.HandleGetValueResult into TryGetTypedValue

File: SignalRServiceBenchmarkPlugin/common/Extensions.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            TValue tmp;
+            if (data.TryGetValue(key, out tmp))
+            {
+                if (converter != null)
+                {
+                    value = converter(tmp);
+                    return true;
+                }
+                if (tmp is TActual)
```
Comment (vwxyzh) [inline SignalRServiceBenchmarkPlugin/common/Extensions.cs]: What about `{ "a": null }.TryGetTypedValue("a", out string x)`?

File: SignalRServiceBenchmarkPlugin/framework/master/Program.cs [hunk 4]
```diff
@@ -70,19 +74,8 @@ private static async Task InstallPluginInSlaves(IList<IRpcClient> clients, strin
...
-
-            // Check whether the plugin installed
-            var success = true;
-            foreach (var task in tasks)
-            {
-                if (!task.Result) success = false;
-                break;
-            }
+            var installResults = await Task.WhenAll(from client in clients select client.InstallPluginAsync(moduleName));
+            var success = installResults.All(result => result == true);
```
Comment (vwxyzh) [inline SignalRServiceBenchmarkPlugin/framework/master/Program.cs]: simplify: `var success = installResults.All(x => x);`

File: SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/MasterMethods/CreateConnection.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+{
+    public class CreateConnection : IMasterMethod
+    {
+        public Task Do(IDictionary<string, object> stepParameters, IDictionary<string, object> pluginParameters, IList<IRpcClient> clients)
+        {
+
+            Log.Information($"Create connections...");
+
+            // Get parameters
+            PluginUtils.TryGetTypedValue(stepParameters, SignalRConstants.ConnectionTotal, out int connectionTotal, Convert.ToInt32);
```
Comment (vwxyzh) [inline SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/MasterMethods/CreateConnection.cs]: `TryGetTypedValue` is extension method

File: SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/SlaveMethods/CreateConnection.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        private IList<HubConnection> CreateConnections(int total, string url, string transportTypeString, string protocolString)
+        {
+            var success = true;
+
+            success = Enum.TryParse<HttpTransportType>(transportTypeString, true, out var transportType);
+            PluginUtils.HandleParseEnumResult(success, transportTypeString);
+
+            var connections = Enumerable.Repeat<HubConnection>(null, total);
+            connections =
+            from connection in connections
```
Comment (vwxyzh) [inline SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/SlaveMethods/CreateConnection.cs]: indent

File: SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/SlaveMethods/CreateConnection.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        private IList<HubConnection> CreateConnections(int total, string url, string transportTypeString, string protocolString)
+        {
+            var success = true;
+
+            success = Enum.TryParse<HttpTransportType>(transportTypeString, true, out var transportType);
+            PluginUtils.HandleParseEnumResult(success, transportTypeString);
+
+            var connections = Enumerable.Repeat<HubConnection>(null, total);
```
Comment (vwxyzh) [inline SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/SlaveMethods/CreateConnection.cs]: `Enumerable.Range`

File: SignalRServiceBenchmarkPlugin/plugins/base/PluginInterface/ISlavehandler.cs [hunk 8]
```diff
@@ -1,10 +1,14 @@
-﻿using System;
+﻿using Rpc.Service;
+using System;
 using System.Collections.Generic;
 using System.Text;
+using System.Threading.Tasks;
 
 namespace Plugin.Base
 {
     public interface ISlavehandler
```
Comment (vwxyzh) [inline SignalRServiceBenchmarkPlugin/plugins/base/PluginInterface/ISlavehandler.cs]: `ISlaveHandler`

File: SignalRServiceBenchmarkPlugin/plugins/base/PluginInterface/IMasterMethod.cs [hunk 9]
```diff
@@ -0,0 +1,12 @@
+﻿using Rpc.Service;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Plugin.Base
+{
+    public interface IMasterMethod
```
Comment (vwxyzh) [inline SignalRServiceBenchmarkPlugin/plugins/base/PluginInterface/IMasterMethod.cs]: naming, what about `Command`

File: SignalRServiceBenchmarkPlugin/framework/master/Program.cs (hunk 4 above)
Comment (realwanpengli) [reply SignalRServiceBenchmarkPlugin/framework/master/Program.cs]: Got it, thanks

File: SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/SlaveMethods/CreateConnection.cs (hunk 6 above)
Comment (realwanpengli) [reply SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/SlaveMethods/CreateConnection.cs]: Got it, thanks

File: SignalRServiceBenchmarkPlugin/plugins/base/PluginInterface/ISlavehandler.cs (hunk 8 above)
Comment (realwanpengli) [reply SignalRServiceBenchmarkPlugin/plugins/base/PluginInterface/ISlavehandler.cs]: Got it, thanks

File: SignalRServiceBenchmarkPlugin/plugins/base/PluginInterface/IMasterMethod.cs (hunk 9 above)
Comment (realwanpengli) [reply SignalRServiceBenchmarkPlugin/plugins/base/PluginInterface/IMasterMethod.cs]: `Command` is more make sense, thanks

File: SignalRServiceBenchmarkPlugin/common/Extensions.cs (hunk 3 above)
Comment (realwanpengli) [reply SignalRServiceBenchmarkPlugin/common/Extensions.cs]: This extension may have several edge cases, will be covered or narrow down the coverage in next PR.

File: SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/SlaveMethods/CreateConnection.cs (hunk 7 above)
Comment (realwanpengli) [reply SignalRServiceBenchmarkPlugin/plugins/Plugin.Microsoft.Azure.SignalR.Benchmark/SignalRPlugin/SlaveMethods/CreateConnection.cs]: your code snippet is more readable, thanks.

**Pull request merged**